src/base/listener.ts

```diff
@@ -477,29 +464,73 @@ class Listener extends EventEmitter implements InterfaceListener {
    * Binds the process to a TCP socket
    * @param opts host and port to bind to
    */
-  private async listenTCP(opts?: { host: string; port: number }): Promise<number> {
-    await new Promise<void>((resolve, reject) => {
-      this.tcpSocket.once('error', (err: any) => {
-        this.tcpSocket.removeListener('listening', resolve)
-        reject(err)
-      })
+  private async listenTCP(opts?: { host?: string; port: number }): Promise<number> {
+    await this.bindToPort('TCP', opts)
 
-      this.tcpSocket.once('listening', () => {
-        this.tcpSocket.removeListener('error', reject)
-        resolve()
-      })
+    return (this.tcpSocket.address() as AddressInfo).port
+  }
+
+  private bindToPort(protocol: 'UDP' | 'TCP', opts?: { host?: string; port: number }) {
+    return new Promise<void>((resolve, reject) => {
+      let socket: TCPServer | UDPSocket
+      let done = false
+
+      switch (protocol) {
+        case 'TCP':
+          socket = this.tcpSocket
+          break
+        case 'UDP':
+          socket = this.udpSocket
+          break
+        default:
+          throw Error()
+      }
+
+      const errListener = (err: any) => {
+        socket.removeListener('listening', successListener)
+        if (!done) {
+          done = true
+          reject(err)
+        }
+      }
+
+      const successListener = () => {
+        socket.removeListener('error', errListener)
+        if (!done) {
+          done = true
+          resolve()
+        }
+      }
+
+      socket.once('error', errListener)
+      socket.once('listening', successListener)
 
       try {
-        this.tcpSocket.listen(opts)
+        switch (protocol) {
+          case 'TCP':
+            ;(socket as TCPServer).listen(opts)
```

Comment:
what do these semicolons do?

Reply:
coming from Prettier since the code can be misinterpreted as function calls

semicolons seem to reduce the ambiguity in here